[svg] make `d` a presentation attribute

graouts opened the pull request:
#### 5187e475ef752de5bc38ae9570c26e03b2edbe14
<pre>
[svg] make `d` a presentation attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=272509">https://bugs.webkit.org/show_bug.cgi?id=272509</a>

Reviewed by Said Abou-Hallawa.

We now make the `d` SVG attribute map to the `d` CSS property as part of the presentation
attribute system by adding the required mapping under `SVGElement::cssPropertyIdForSVGAttributeName()`
and accounting for the path data set on the `SVGRenderStyle` in `pathFromPathElement()`.

We must also call `setPresentationalHintStyleIsDirty()` under `SVGPathElement::svgAttributeChanged()`,
matching the behavior of other SVG elements with presentation attributes.

The SVG presentation attribute system forwards the string value set on the SVG attribute to the
`MutableStyleProperties` object collecting the various matching CSS properties. But in the case
of the `d` property, path data can be large and unwieldy, so we override `collectPresentationalHintsForAttribute()`
on `SVGPathElement` and create a `CSSPathValue` with a copy of the (potentially animated)
`SVGPathByteStream` and call into a new variant of `StyledElement::addPropertyToPresentationalHintStyle()`
that takes in a `RefPtr&lt;CSSValue&gt;&amp;&amp;`.

Finally, adding support for `d` as a presentation attribute revealed a mistake in 277297@main
where we accounted for the CSS `zoom` value when converting the `CSSPathValue`, but the `zoom`
property has no bearing in SVG. This broke the following tests:

    * svg/zoom/page/zoom-coords-viewattr-01-b.svg
    * svg/zoom/page/zoom-img-preserveAspectRatio-support-1.html
    * svg/zoom/page/zoom-replaced-intrinsic-ratio-001.htm

So we modified `BuilderConverter::convertSVGPath()` accordingly.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/getComputedStyle-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/test_style_flush_on_dom_api_with_d_property-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-relevant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-special-cases-expected.txt:
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::addPropertyToPresentationalHintStyle):
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/rendering/style/BasicShapes.h:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
(WebCore::SVGRenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSVGPath):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::cssPropertyIdForSVGAttributeName):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::svgAttributeChanged):
(WebCore::SVGPathElement::path const):
(WebCore::SVGPathElement::collectPresentationalHintsForAttribute):
* Source/WebCore/svg/SVGPathElement.h:

Canonical link: <a href="https://commits.webkit.org/277451@main">https://commits.webkit.org/277451@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d8c7e4286a2fd9671a6e250464b7b16034a0e47

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47608 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50290 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43657 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49915 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32580 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24258 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38762 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48190 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24414 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41023 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20061 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21901 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42211 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5650 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43951 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42628 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52182 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22644 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18974 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46074 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23916 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45103 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24704 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6732 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6981,14 +6981,21 @@ static RefPtr<CSSPathValue> consumeBasicShapePath(CSSParserTokenRange& args, Opt
-    if (!buildSVGPathByteStreamFromString(args.consumeIncludingWhitespace().value(), byteStream, UnalteredParsing) || byteStream.isEmpty())
+    switch (args.peek().type()) {
+    case IdentToken:
+        return pathValueFromStringView(args.consumeAll().serialize());
```
Do we want to `consumeAll()` here or just the current ident token (especially when the string token case only consume a single token) ?

graouts replied:
I believe we do because there will be multiple tokens to match the long list of path data commands.

mdubet replied:
Isn't the case of multiple data commands handle by the `StringToken` case ? Can you have multiple command written as IdentToken (so without string delimiters) ?

graouts replied:
Yes, when the value comes internally from the SVG attribute parsing code. See this code:
```
void SVGElement::collectPresentationalHintsForAttribute(const QualifiedName& name, const AtomString& value, MutableStyleProperties& style)
{
    CSSPropertyID propertyID = cssPropertyIdForSVGAttributeName(name);
    if (propertyID > 0)
        addPropertyToPresentationalHintStyle(style, propertyID, value);
}
```
The `value` here is coming straight from the SVG attribute.

graouts replied:
This code has been entirely removed since we now bypass style resolution altogether and manually create a `CSSPathValue` with the `SVGPathByteStream`.

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6981,14 +6981,21 @@ static RefPtr<CSSPathValue> consumeBasicShapePath(CSSParserTokenRange& args, Opt
+    case IdentToken:
+        return pathValueFromStringView(args.consumeAll().serialize());
+    case StringToken:
+        return pathValueFromStringView(args.consumeIncludingWhitespace().value());
```
Should we verify that they are no token pending ? 

```
if (!args.atEnd())
      return nullptr;
```

graouts replied:
The call sites do. I'm not sure whether we should factor this into `consumeBasicShapePath()` or let the call sites handle it. See for instance `consumeBasicShape()` which has this single check at its end:

```
    if (!result || !args.atEnd())
        return nullptr;
```

graouts replied:
This code has been removed.

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6981,14 +6981,21 @@ static RefPtr<CSSPathValue> consumeBasicShapePath(CSSParserTokenRange& args, Opt
 
-    if (args.peek().type() != StringToken)
-        return nullptr;
+    auto pathValueFromStringView = [&](const StringView& view) -> RefPtr<CSSPathValue> {
```
We should have a guideline about `StringView` and `const StringView&` but I don't think we do 😬

mdubet replied:
Using StringView by value is preferred

graouts replied:
Will fix in the updated commit.

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7228,15 +7235,13 @@ RefPtr<CSSValue> consumePathOperation(CSSParserTokenRange& range, const CSSParse
-    if (!result || !args.atEnd())
-        return nullptr;
-    return result;
+    if (range.peek().type() == FunctionToken && range.peek().functionId() == CSSValuePath) {
```
We generally favour early return for error (like the old code)

graouts replied:
Alright, I've rewritten it to use early returns in the latest patch.

graouts replied:
This code has been removed.

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -217,4 +219,14 @@ RenderPtr<RenderElement> SVGPathElement::createElementRenderer(RenderStyle&& sty
     return createRenderer<LegacyRenderSVGPath>(*this, WTFMove(style));
 }
 
+Path SVGPathElement::path() const
+{
+    if (auto* renderer = this->renderer()) {
+        if (auto* basicShapePath = renderer->style().d())
+            return basicShapePath->path({ });
+    }
+
+    return m_pathSegList->currentPath();
+}
```
Can you explain what is this function returning? Are we allowing the `m_pathSegList` to be out of sync from the style property? What should JS return in case of SMIL and CSS animations? What is the expected value of
```
pathSegList
animatedPathSegList
getAttribute('d')
```

graouts replied:
I made a [little test](https://codepen.io/graouts/pen/BaEVOBe) for this, dumping the various value half-way through an animation performed by either an SVG `<animate>` element or a CSS Animation from `M 0 0 H 100 V 100 H 0 Z` to `M 0 0 H 200 V 200 H 0 Z` (values rounded for legibility):

```
Path animated with SVG
    pathSegList = M 0 0 H 100 V 100 H 0 z
    animatedPathSegList = M 0 0 H 150 V 150 H 0 z
    getAttribute("d") = M 0 0 H 100 V 100 H 0 Z
    computed style = path("M 0 0 H 150 V 150 H 0 Z")

Path animated with CSS
    pathSegList = M 0 0 H 100 V 100 H 0 z
    animatedPathSegList = M 0 0 H 100 V 100 H 0 z
    getAttribute("d") = M 0 0 H 100 V 100 H 0 Z
    computed style = path("M 0 0 H 150 V 150 H 0 Z")
```

The `getAttribute("d")` and computed style for the `d` property match what I see in Chrome Canary and Firefox Nightly. Both of those browsers having dropped support for `pathSegList` and `animatedPathSegList` I cannot compare those values.

This all makes sense to me.

shallawa replied:
I think this makes sense to me also.

I think I was confused by the fact this function is in `SVGPathElement`. I think getting the `Path` from a `SVGPathElement` is a rendering functionality. Maybe later `SVGPathElement::path()` is be removed entirely and its code is moved to `pathFromPathElement()`.

pathSegList, animatedPathSegList and getAttribute() should not be affected by the CSS animation.

graouts replied:
Note that I did not introduce this function, I just moved its implementation in the `.cpp` file since it now had more code.

shallawa reviewed (approved):
Since perf/svg-path-appenditem.html was failing at some point, it may be worthy mentioning this test was added in https://bugs.webkit.org/show_bug.cgi?id=94048 because of this benchmark http://bl.ocks.org/1296930. This benchmark runs only in Safari right now.

graouts committed:
[svg] make `d` a presentation attribute
https://bugs.webkit.org/show_bug.cgi?id=272509

Reviewed by Said Abou-Hallawa.

We now make the `d` SVG attribute map to the `d` CSS property as part of the presentation
attribute system by adding the required mapping under `SVGElement::cssPropertyIdForSVGAttributeName()`
and accounting for the path data set on the `SVGRenderStyle` in `pathFromPathElement()`.

We must also call `setPresentationalHintStyleIsDirty()` under `SVGPathElement::svgAttributeChanged()`,
matching the behavior of other SVG elements with presentation attributes.

The SVG presentation attribute system forwards the string value set on the SVG attribute to the
`MutableStyleProperties` object collecting the various matching CSS properties. But in the case
of the `d` property, path data can be large and unwieldy, so we override `collectPresentationalHintsForAttribute()`
on `SVGPathElement` and create a `CSSPathValue` with a copy of the (potentially animated)
`SVGPathByteStream` and call into a new variant of `StyledElement::addPropertyToPresentationalHintStyle()`
that takes in a `RefPtr<CSSValue>&&`.

Finally, adding support for `d` as a presentation attribute revealed a mistake in 277297@main
where we accounted for the CSS `zoom` value when converting the `CSSPathValue`, but the `zoom`
property has no bearing in SVG. This broke the following tests:

    * svg/zoom/page/zoom-coords-viewattr-01-b.svg
    * svg/zoom/page/zoom-img-preserveAspectRatio-support-1.html
    * svg/zoom/page/zoom-replaced-intrinsic-ratio-001.htm

So we modified `BuilderConverter::convertSVGPath()` accordingly.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/getComputedStyle-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/test_style_flush_on_dom_api_with_d_property-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-relevant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-special-cases-expected.txt:
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::addPropertyToPresentationalHintStyle):
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/rendering/style/BasicShapes.h:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
(WebCore::SVGRenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSVGPath):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::cssPropertyIdForSVGAttributeName):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::svgAttributeChanged):
(WebCore::SVGPathElement::path const):
(WebCore::SVGPathElement::collectPresentationalHintsForAttribute):
* Source/WebCore/svg/SVGPathElement.h:

Canonical link: https://commits.webkit.org/277451@main
